### PR TITLE
Symbolic expressions now support delayed binding to arguments. Fixes #93.

### LINKED ISSUE #93: Type checking on closures
Hey @patrick-kidger,
great package as always. I was wondering what the best way is to type check closure functions? For example:

```python
import numpy as np
from jaxtyping import Float
from typeguard import typechecked


def f(n):
    @typechecked
    def g(x: Float[np.ndarray, f"dim1 {n}"]) -> Float[np.ndarray, f"{n}"]:
        return x.sum(0)

    return g


print(f(10)(np.random.randn(10, 9)))
```

This seems to work, but I'm wondering if this is the proper way? Or maybe something with `typing.TypeVar` is better?

I should also mention I'm not necessarily interested in closures, but more in class methods, where I want the return type to depend on a class parameter (e.g., `-> Float[Array, f"{self.latent}"]`).

### PR DESCRIPTION
**Pull request by patrick-kidger**

### CODE REVIEW

File: docs/api/array.md [hunk 1]
```diff
@@ -37,9 +38,18 @@ As a special case:
 - To denote a scalar shape use `""`, e.g. `Float[Array, ""]`.
 - To denote an arbitrary shape (and only check dtype) use `"..."`, e.g. `Float[Array, "..."]`.
 - You cannot have more than one use of multiple-axes, i.e. you can only use `...` or `*name` at most once in each array.
-- An example of broadcasting multiple axes:  
-    `def add(x: Float[Array, "*#foo"], y: Float[Array, "*#foo"]) -> Float[Array, "*#foo"]`.
 - A symbolic expression cannot be evaluated unless all of the axes sizes it refers to have already been processed. In practice this usually means that they should only be used in annotations for the return type, and only use axes declared in the arguments.
+- Symbolic expressions are evaluated in two stages: they are first evaluated as f-strings using the arguments of the function, and second are evaluated using the processed axis sizes. The f-string evaluation means that they can use local variables by enclosing them with curly braces, e.g. `{variable}`, e.g.
+    ```python
+    def full(size: int, fill: float) -> Float[Array, "{shape}"]:
```
Comment (dlwh) [inline docs/api/array.md:44]: pretty sure this should be `{size}`?

Comment (patrick-kidger) [reply docs/api/array.md:44]: Agreed! I'll fix it now.
This does make me think, though -- handling a general `shape` is a little annoying with this change:
```python
def full(shape: tuple[int, ...], fill: float) -> Float[Array, "{strshape}"]:
    strshape = " ".join(map(str, shape))
    return jnp.full(shape, fill)
```
I don't have a good answer for that right now.

Comment (dlwh) [reply docs/api/array.md:44]: yeah i thought about this for a minute and decided it was tough. maybe just a helper function? `jaxtyping.strshape` that just does the first line? tuple shapes are pretty common!

**Comment by danielward27** [comment]
Not sure if this is intended, but after reading the updated docs it isn't obvious to me why one of these would work, and the other would not
```python

import equinox as eqx
import jax.numpy as jnp
from beartype import beartype as typechecker
from jaxtyping import ArrayLike, Shaped, jaxtyped


class SomeClass1(eqx.Module):
    dim: int

    @jaxtyped
    @typechecker
    def __call__(self, arr: Shaped[ArrayLike, "2 {self.dim}"]):
        return arr


class SomeClass2(eqx.Module):
    dim: int

    @jaxtyped
    @typechecker
    def __call__(self, arr: Shaped[ArrayLike, "{self._strshape}"]):
        return arr

    @property
    def _strshape(self):
        return f"2 {self.dim}"


dim = 5
x = jnp.ones((2, dim))
my_class1 = SomeClass1(dim)
my_class2 = SomeClass2(dim)


my_class1(x)  # This works
my_class2(x)  # This raises BeartypeCallHintParamViolation

```

**Comment by patrick-kidger** [comment]
Ah, that's an interesting edge-case. The reason is that delayed bindings of the form `"{self.foo}"` can only be used for a single name (whether that's a single axis e.g. `batch` or variadic axes e.g. `*batch`). Here you're trying to expand to two names: `2 5`.

I suspect that's probably a limitation that can be lifted though! I'll see what I can do.

**Comment by patrick-kidger** [comment]
Having now looked at this a bit, I don't think this is something that can easily be arranged. Right now the code for separating out each axis runs when the annotation is first created. Meanwhile, the code for these kinds of f-strings is ran during the `isinstance` check.

So in order to support use-cases like the above, we would need to delay running the "axis handling" code until after the f-string handling code. This would (a) represent a small performance drop for unJIT'd code, and (b) be a fairly fiddly rewrite to arrange.

As such I think this probably won't be something supported, at least in the near future.

**Comment by danielward27** [comment]
I see, that seems reasonable. Thanks for having a look anyway! I guess this means objects with ``tuple[int, ...]`` shape attributes can't use their shapes in the parameter annotations too (unless you know the length of the shape beforehand, in which case you could manually splat it out).
